### DISCUSSION TITLE
Unable to get Micro SD Card to work

### DISCUSSION THREAD

**Discussion by Slyke** [arendst/Tasmota#15236] [Q&A] [answered]
I have a DS1307 RTC SD Card board and have connected it to my NodeMCU v3 running Tasmota 11.0.0. I have the pins connected and setup as follows:
![image](https://user-images.githubusercontent.com/6442613/160275458-c8a8547b-489b-4644-ae4b-e0116cda4b8d.png)

I have also tried switching the SDCard CS to GPIO9 in configuration (and plugging it into the S2 pin on the NodeMCU).

On power up, I get the following log:
```
00:00:00.065 SPI: Hardware using GPIO14(CLK), GPIO13(MOSI) and GPIO12(MISO)
00:00:00.269 Project tasmota - Portable Sensor Logger Prototype Version 11.0.0(sensors)-2_7_4_9(2022-03-26T17:23:21)
00:00:00.270 UFILESYSTEM fail, using compression!
```

Trying out some of the UFS commands:
```
10:26:26.262 CMD: UfsFree
10:26:26.267 RSL: RESULT = {"UfsFree":0}
10:26:46.497 CMD: UfsSize
10:26:46.502 RSL: RESULT = {"UfsSize":0}
10:26:52.921 CMD: UfsType
10:26:52.927 RSL: RESULT = {"UfsType":0}
```

I have compiled the tasmota-sensors build with:
```
#ifndef USE_GPS
#define USE_GPSbytes RAM)
  #define USE_FLOG
#endif

#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
#define SDCARD_DIR
#define USE_SCRIPT_FATFS_EXT
#define USE_SCRIPT_FATFS
#define USE_UFILESYS
#define USE_SDCARD
#define SDCARD_CS_PIN 4
```

Which I got from this post: https://github.com/arendst/Tasmota/discussions/13938

The problem is that on the Consoles page I get the following:
![image](https://user-images.githubusercontent.com/6442613/160275571-c3b3d90b-f84f-40c4-9683-08ea375bc334.png)
and if I click "Edit Script":
![image](https://user-images.githubusercontent.com/6442613/160275585-ef577450-f6fc-427f-8734-fc39a2ce5063.png)

I'm not sure what I'm doing wrong. I can't seem to get the file system option to show up.

I'm using a known good micro sd card and have formatted it to FAT.

**Comment by pkkrusty** [comment]
As far as I know, you compile custom builds like that with the base tasmota.bin, not one of the extras, like tasmota-sensors. I see you define GPIO4 for the Chip Select pin, but your config shows you choosing a different pin. I noticed that for some reason, I had to define the SPI pins exactly as shown in my post.  If I chose other ones, it wouldn't work. 

Also make sure you have `#undef USE_RULES` declared since you declared scripting. 

Try adjusting those things and see if it works.

**Reply by Slyke** [reply]
> I see you define GPIO4 for the Chip Select pin

This is probably the issue. Is there a way to workout which pin is for which GPIO without trial and error? I did find this diagram
![image](https://user-images.githubusercontent.com/6442613/160276880-445aabbe-fc6c-4e08-bbd1-623f63df6ac5.png)

And have tried setting #define SDCARD_CS_PIN to 4 and 9, swapping the wire, and updating the config, but still same result. I'll keep messing with it.

> Also make sure you have `#undef USE_RULES` declared since you declared scripting

Done! I did have it commented out in the `my_user_config.h` file already, but added it back there since it's now being undefined.

**Comment by pkkrusty** [comment]
Looks like a good diagram for it.  So you connected SDCARD CS to D2 when defined as GPIO 4?

The module you're using looks like a hat. Are you connected with wires? And confirm that you're powering the module with 5v

**Reply by Slyke** [reply]
Yep, have tried others too while recompiling it for testing other pins

**Comment by gemu2015** [comment]
you must enable FLASH file system also. SD as file system alone is currently not supported.

like @pkkrusty already mentioned in his post:

In platform_override.ini, I commented out
;board = ${common.board}
and uncommented
board = esp8266_4M2M
to have the option of using the extra 2mb flash on the Wemos knockoff chip I'm using.

**Reply by pkkrusty** [reply]
Didn't realize that was actually necessary. Guess I got lucky.

**Reply by Slyke** [reply]
Seems that has made it kind of work. It's detecting the flash filesystem on the NodeMCU, but still not detecting the SD Card:

```
00:00:00.001 HDW: ESP8266EX
00:00:00.011 UFS: FlashFS mounted with 1992 kB free
00:00:00.071 CFG: Loaded from File, Count 107
00:00:00.076 FRC: Some settings have been reset (5)
00:00:00.086 Project tasmota - Portable Sensor Logger Prototype Version 11.0.0(sensors)-2_7_4_9(2022-03-27T21:58:41)
00:00:00.087 UFILESYSTEM OK!
...
06:16:43.067 RSL: INFO3 = {"Info3":{"RestartReason":{"Exception":29,"Reason":"Exception","EPC":["40271f3c","00000000","00000000"],"EXCVADDR":"00000000","DEPC":"00000000","CallChain":["402575f9","40101494","40257935","402579ba","40257a94","402363eb","4023653a","40236562","4023662c","4023666a","40265f74","40237c84","402a854b","402389e2","40101ba3","40101e27","40101ba3","40101e27","402a9102","40238b54","40248f09","402578b6","40257e89","40266604","4022a9c8","4026d0ca","4026d128","401000e1","4022aa00","402660ad","4023564e"]},"BootCount":59}}
06:16:46.303 QPC: Reset
06:16:47.263 RSL: STATE = {"Time":"2022-03-28T06:16:47","Uptime":"0T00:00:09","UptimeSec":9,"Heap":12,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Wifi":{"AP":1,"SSId":"TestNet","BSSId":"78:D2:94:1A:05:27","Channel":8,"Mode":"11n","RSSI":100,"Signal":-46,"LinkCount":1,"Downtime":"0T00:00:03"}}
06:17:06.888 CMD: ufstype
06:17:06.894 RSL: RESULT = {"UfsType":3}
```

I Notice that it's very slow on the webUI. I suspect it's rebooting due to Exception 29 listed above.

**Reply by Slyke** [reply]
Still can't get it to detect the SD Card though, not sure if it's a config issue or how to troubleshoot further.

I've tried pins D0, D2, D3, D4, S2 and S3 for SDCARD_CS_PIN and none seem to get it to work. Been updating the config too, but no luck.

**Comment by gemu2015** [comment]
#define SDCARD_CS_PIN is not needed at all. it is ignored if you select the pin with the GPIO web UI.
anyhow you should not get exception 29.  as long as you get this something else is wrong
using a predefined build like sensors may introduce problems.

when testing do only define SD card pins, nothing more.  GPIO2 for SDCARD_CS should be fine.
you can add other functionality later

the card must be formatted FAT32
the cabling to the card socket must be as short as possible.
double check the wiring assignment

**Reply by Slyke** [reply]
![image](https://user-images.githubusercontent.com/6442613/160356354-43f8a51d-aff5-41b7-a5c1-31b5e768d789.png)

Double checked all the wiring, disconnected everything else. The wires are using a breadboard, but it's no longer than around 10cm in total.  Formatted the SD Card to FAT32 with Windows, but it still doesn't show up. Going to order another one in case it's faulty, but are there any other troubleshooting steps that could get Tasmota to show what's wrong?

**Comment by gemu2015** [comment]
in your log above i miss the SPI Message

this is the log of an esp8266 with SD card

00:00:00.002 HDW: ESP8266EX
00:00:00.033 UFS: FlashFS mounted with 1920 kB free
00:00:00.095 CFG: Loaded from File, Anzahl 30
00:00:00.101 QPC: Count 1
00:00:00.104 SPI: Hardware using GPIO14(CLK), GPIO13(MOSI) and GPIO12(MISO)
00:00:00.576 UFS: SDCard mounted
00:00:00.580 Projekt tasmota - wemos-d1-mini-sdc Version 2022.01.3(tasmota)-2_7_4_9(2022.02.02 11:12:36)
00:00:00.580 UFILESYSTEM OK!

**Reply by pkkrusty** [reply]
One of his first log postings show the SPI message, but you're right, that latest one doesn't show it. Did #use_spi fall off the build on accident?

**Reply by Slyke** [reply]
I'll throw in this preprocessor and see if it helps. Gonna get a basic Arduino sketch and flash it to the NodeMCU and see if the SDcard works that way.

**Reply by Slyke** [reply]
Seems it won't even work with a basic Arduino sketch. Going to wait until the new boards come in and give it a try then.


```
#include <Arduino.h>
#include <SPI.h>
#include <SD.h>
File myFile;
void setup() {
  Serial.begin(9600);
  while (!Serial) {
    ; // wait for serial port to connect. Needed for native USB port only
  }

  Serial.print("Initializing SD card...");

  if (!SD.begin(D2)) {
    Serial.println("initialization failed!");
    while (1);
  }

  Serial.println("initialization done.");
  myFile = SD.open("test.txt", FILE_WRITE);

  if (myFile) {
    Serial.print("Writing to test.txt...");
    myFile.println("This is a test file :)");
    myFile.println("testing 1, 2, 3.");
    for (int i = 0; i < 20; i++) {
      myFile.println(i);
    }
    myFile.close();
    Serial.println("done.");
  } else {
    Serial.println("error opening test.txt");
  }
}

void loop() {
// nothing happens after setup
}
```

Result:
```
��Initializing SD card...initialization failed!
```

Tried D1, D2, D4 and D8.

Will report back once they've arrived.

**Comment by Jason2866** [comment]
What size does your SD Card have? Afair there is a max. size.

**Reply by Slyke** [reply]
It's 8gb.

**Reply by Jason2866** [reply]
8 Gb does work

**Comment by Slyke** [comment, marked as answer]
Okay, figured out the SDCard issue after a LOT of digging around and trying various hardware things. Apparently the library for NodeMCU is touchy when it comes to the SD Card. Even though it was FAT32, the library couldn't handle it for some reason. It looks like the developer has a new v2 beta version of the library ( https://github.com/greiman/SdFat ).

Formatting with just Windows was not enough.

I formatted the card using this tool: https://www.sdcard.org/downloads/formatter/ and everything started working immediately.

I figured this out by running this code on PlatformIO:

```
#include <Arduino.h>
#include <SPI.h>
#include <SD.h>
#include <SdFat.h>

File myFile;
void setup() {
  Serial.begin(9600);
  while (!Serial) {
    ; // wait for serial port to connect. Needed for native USB port only
  }

  Serial.print("Initializing SD card...");
  delay(250);

  bool initFailed = false;

  if (!SD.begin(D4, SD_SCK_MHZ(1))) {
    initFailed = true;
    Serial.println("initialization failed!");
    Serial.println();
  }
  delay(500);

  Serial.print("Card type:         ");
  Serial.println(SD.type());
  Serial.print("fatType:           ");
  Serial.println(SD.fatType());
  Serial.print("size:              ");
  Serial.println(SD.size());


  Serial.println();
  if (initFailed) {
    delay(1000);
    while (1); // Soft reset
  }

  Serial.println("initialization done.");
  myFile = SD.open("test.txt", FILE_WRITE);

  if (myFile) {
    Serial.print("Writing to test.txt...");
    myFile.println("This is a test file :)");
    myFile.println("testing 1, 2, 3.");
    for (int i = 0; i < 20; i++) {
      myFile.println(i);
    }
    myFile.close();
    Serial.println("done.");
  } else {
    Serial.println("error opening test.txt");
  }
}
void loop() {
// nothing happens after setup
}
```

And noticed that the Card Type was 3, but it had no size being reported.

![image](https://user-images.githubusercontent.com/6442613/162648835-399402a0-bb6f-43fd-9969-a9ced96050ca.png)

@ascillato2 could you please update the selected answer?

**Reply by ascillato2** [reply]
👍  Thanks.
